vyper/semantics/analysis/local.py

```diff
@@ -203,9 +204,22 @@ def __init__(
             ]
             node_list.extend(standalone_self)  # type: ignore
 
+            # Add all calls except structs
+            filtered_call_nodes = [
+                c.func
+                for c in fn_node.get_descendants(vy_ast.Call)
+                if not (len(c.args) == 1 and isinstance(c.args[0], vy_ast.Dict))
+            ]
+            node_list.extend(filtered_call_nodes)  # type: ignore
+
             for node in node_list:
                 t = node._metadata.get("type")
-                if isinstance(t, ContractFunctionT) and t.mutability == StateMutability.PURE:
+                # skip structs and interface instantiation
+                if isinstance(t, StructT) or is_type_t(t, InterfaceT):
+                    continue
+
+                # skip ContractFunctionT and BuiltinFunction with mutability set to pure
+                if hasattr(t, "mutability") and t.mutability == StateMutability.PURE:
```

Comment:
i think the ast analysis is still not the way to go. i think probably a better approach is to add a `mutability` to all `ExprInfo`s, and then during `local.py` we can check at every single expression (and maybe statement too) whether `expr_info.mutability > self.func.mutability`.